CwdUtils.c

```diff
@@ -0,0 +1,163 @@
+#include "config.h" // IWYU pragma: keep
+
+#include "CwdUtils.h"
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <wchar.h>
+
+#include "XUtils.h"
+
+typedef struct ShortenCwdContext {
+   size_t maxLength;
+   size_t len;
+   wchar_t **parts;
+   size_t partsLen;
+   size_t *partLengths;
+} ShortenCwdContext;
+
+void ShortenCwdContext_init(ShortenCwdContext *ctx, wchar_t *cwd, size_t maxLength) {
+   ctx->maxLength = maxLength;
+   ctx->len = wcslen(cwd);
+   ctx->parts = Wstring_split(cwd, L'/', &ctx->partsLen);
+   ctx->partLengths = xCalloc(ctx->partsLen, sizeof(size_t));
+   for (size_t i = 0; i < ctx->partsLen; i++)
+      ctx->partLengths[i] = wcslen(ctx->parts[i]);
+}
+
+void ShortenCwdContext_destroy(ShortenCwdContext *ctx) {
+   for (size_t i = 0; i < ctx->partsLen; i++)
+      free(ctx->parts[i]);
+   free(ctx->parts);
+   free(ctx->partLengths);
+   *ctx = (ShortenCwdContext){0};
+}
+
+static void shortenCwdParts(ShortenCwdContext *ctx) {
+   for (size_t i = ctx->partsLen - 2; i != (size_t)-1 && ctx->len > ctx->maxLength; i--) {
+      if (ctx->partLengths[i] < 3)
```

Comment:
```suggestion
      if (ctx->partLengths[i] <= 3)
```

Also, documenting where the 3 comes from (initial, separator, trailer) would be nice.